src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/common/actions/YFlowSpeakerProcessingAction.java

```diff
@@ -0,0 +1,131 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.UUID;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+@Slf4j
+public abstract class YFlowSpeakerProcessingAction<T extends YFlowProcessingFsm<T, S, E, C, ?, ?>, S, E, C>
+        extends YFlowProcessingWithHistorySupportAction<T, S, E, C> {
+    protected final RuleManager ruleManager;
+
+    public YFlowSpeakerProcessingAction(PersistenceManager persistenceManager, RuleManager ruleManager) {
+        super(persistenceManager);
+        this.ruleManager = ruleManager;
+    }
+
+    protected Collection<InstallSpeakerCommandsRequest> buildYFlowInstallCommands(YFlow yFlow, CommandContext context) {
+        Map<SwitchId, List<SpeakerData>> speakerData = buildYFlowSpeakerData(yFlow);
+
+        return speakerData.entrySet().stream().map(entry -> {
+            SwitchId switchId = entry.getKey();
+            List<SpeakerData> dataList = entry.getValue();
+            UUID commandId = commandIdGenerator.generate();
+            MessageContext messageContext = new MessageContext(commandId.toString(), context.getCorrelationId());
+            return new InstallSpeakerCommandsRequest(messageContext, switchId, commandId, mapToOfCommands(dataList));
+        }).collect(Collectors.toList());
+    }
+
+    protected Collection<DeleteSpeakerCommandsRequest> buildYFlowDeleteCommands(YFlow yFlow, CommandContext context) {
+        Map<SwitchId, List<SpeakerData>> speakerData = buildYFlowSpeakerData(yFlow);
+
+        return speakerData.entrySet().stream().map(entry -> {
+            SwitchId switchId = entry.getKey();
+            List<SpeakerData> dataList = entry.getValue();
+            UUID commandId = commandIdGenerator.generate();
+            MessageContext messageContext = new MessageContext(commandId.toString(), context.getCorrelationId());
+            return new DeleteSpeakerCommandsRequest(messageContext, switchId, commandId, mapToOfCommands(dataList));
+        }).collect(Collectors.toList());
+    }
+
+    private Map<SwitchId, List<SpeakerData>> buildYFlowSpeakerData(YFlow yFlow) {
+        List<FlowPath> flowPaths = yFlow.getSubFlows().stream()
+                .map(YSubFlow::getFlow)
+                .map(Flow::getPaths)
+                .flatMap(Collection::stream)
+                .collect(toList());
+
+        Set<SwitchId> switchIds = Sets.newHashSet(yFlow.getSharedEndpoint().getSwitchId(), yFlow.getYPoint(),
+                yFlow.getProtectedPathYPoint());
+        Set<PathId> pathIds = flowPaths.stream()
+                .map(FlowPath::getPathId)
+                .collect(Collectors.toSet());
+        DataAdapter dataAdapter = PersistenceDataAdapter.builder()
+                .persistenceManager(persistenceManager)
+                .switchIds(switchIds)
+                .pathIds(pathIds)
+                .build();
+
+        return ruleManager.buildRulesForYFlow(flowPaths, dataAdapter).stream()
```

Comment:
So, it looks like you convert the y-flow into a set of flow paths, and then inside of RM you convert them back. Is it as expected?

Reply:
At first I did not pick out the paths, but due to the fact that we are using adapters, it turned out that as I did it easier.